Miscellaneous fixes for confined users

gucci-on-fleek opened the pull request:
I've been running using SELinux-confined users on my Fedora IoT server for a few months now, but I've needed a few custom rules for it to work properly. This patch contains those rules.

There are quite a few (very small) changes here, so please let me know if you have any questions or want me to make any changes.

Full disclosure: I've been using a [CIL version of these rules](https://github.com/gucci-on-fleek/maxchernoff.ca/blob/1d593bb8dfc281845c7efd7215d2da7b51c4eedf/etc/selinux/local-policies/local-systemd.cil) for a couple months now, and I've confirmed that the policies still build (`make all`) and validate (`make validate`) correctly, but I haven't tested installing the `.pp` modules from this patch on my server.

(Vaguely related to containers/container-selinux#383)

gucci-on-fleek committed:
Allow sysadm to communicate with systemd-networkd

Right now, with confined users enabled, running `sudo networkctl status`
(or any other subcommand) fails with permission errors. This commit
makes it possible for a confined administrator to run `networkctl`
commands.

gucci-on-fleek committed:
Silence journalctl AVC denials with lesshst files

gucci-on-fleek committed:
Allow runtime-generated systemd user units for confined users

gucci-on-fleek committed:
Make `sudo systemctl --user --machine=USER@ ...` work for confined users

zpytela commented on policy/modules/system/systemd.if:
```diff
@@ -2843,6 +2843,29 @@ interface(`systemd_map_resolved_exec_files',`
+	allow $1 systemd_networkd_t:dbus send_msg;
+	allow systemd_networkd_t $1:dbus send_msg;
+	allow $1 systemd_networkd_t:unix_stream_socket connectto;
+	ps_process_pattern(systemd_networkd_t, $1)
```
This looks like 2 or 3 different permission sets. Can you elaborate for what is that needed?

gucci-on-fleek replied:
I copied this from `systemd_chat_resolved` almost exactly, except with `systemd_networkd_t` instead of `systemd_resolved_t`. My personal testing showed that 

```cil
(allow sysadm_t systemd_networkd_t (dbus               (send_msg )))
(allow sysadm_t systemd_networkd_t (unix_stream_socket (connectto)))
```

was sufficient to make this work though, so I can reduce the permissions to that if you want.

gucci-on-fleek replied:
Ok, I've narrowed `systemd_chat_resolved` down to only the needed permissions now.

zpytela commented on policy/modules/contrib/snapper.te:
```diff
@@ -98,3 +98,7 @@ optional_policy(`
 ')
+
+optional_policy(`
+    seutil_read_file_contexts(snapperd_t)
```
I am not convinced this should be added to policy.

gucci-on-fleek replied:
It seems to be required if you're using snapper to snapshot `/etc/` with unconfined disabled. Specifically snapshotting `/etc/` isn't very common, but snapshotting all of `/` is (and `/etc/` is included in that). But I can remove this if you want.

gucci-on-fleek replied:
I was wrong above; replacing this with `seutil_dontaudit_read_file_contexts` fixes the problem.

zpytela commented on policy/modules/contrib/journalctl.te:
```diff
@@ -61,3 +61,14 @@ userdom_rw_inherited_user_home_content_files(journalctl_t)
+		type gconf_home_t;
+	')
+
+	dontaudit journalctl_t gconf_home_t:dir { add_name remove_name write };
```
For using types from other modules, interfaces need to be used.

gucci-on-fleek replied:
Ah, I had no idea about that, sorry. I'll fix that for the next push.

gucci-on-fleek replied:
Ok, this should be fixed now.

zpytela commented on policy/modules/system/systemd.fc:
```diff
@@ -126,6 +126,8 @@ HOME_DIR/\.config/systemd/user(/.*)?		gen_context(system_u:object_r:systemd_unit
 /run/systemd/generator		-d	gen_context(system_u:object_r:systemd_unit_file_t,s0)
 /run/systemd/generator/.+	<<none>>
+/run/user/%{USERID}/systemd/generator		-d	gen_context(system_u:object_r:systemd_unit_file_t,s0)
+/run/user/%{USERID}/systemd/generator/.+	<<none>>
```
Why is it needed to label user units, or is it intended to have private types?

gucci-on-fleek replied:
`systemctl --user` can't access/manipulate them otherwise.

zpytela commented on policy/modules/system/systemd.te:
```diff
@@ -237,6 +237,10 @@ files_type(systemd_zram_generator_conf_t)
+# User domains
+allow userdomain systemd_unit_file_t:dir manage_dir_perms;
+allow userdomain systemd_unit_file_t:file manage_file_perms;
+allow userdomain systemd_unit_file_t:lnk_file manage_lnk_file_perms;
```
Types/attributes from other modules need to be accessed via interfaces. Anyway allowing any user type manage systemd_unit_file_t does not look good.

gucci-on-fleek replied:
`~/.config/systemd/user/` is labelled as `systemd_unit_file_t`

https://github.com/fedora-selinux/selinux-policy/blob/66492b704f2e5d365a701ce05b8623764db93aea/policy/modules/system/systemd.fc#L2

so this is needed for confined users to be able to edit the files there with a text editor. So if you don't want for users to be able to edit all `systemd_unit_file_t`s, maybe we should add a new `systemd_user_unit_file_t`?

zpytela commented on policy/modules/system/systemd.te:
```diff
@@ -971,6 +971,15 @@ init_read_state(systemctl_domain)
+		type sysadm_t;
+	')
+
+	allow userdomain sysadm_t:unix_stream_socket { getattr getopt read setopt write };
```
Only staff_t and admindomains are allowed to run sudo. Interface needs to be used.

gucci-on-fleek replied:
> Only staff_t and admindomains are allowed to run sudo.

Right, but this is needed for the confined user session created with `--machine=USER@` to be able to connect back to the `systemctl` command. Or should we maybe create a new label for that socket file so that we don't need to use `sysadm_t`?

gucci-on-fleek replied:
Ok, I tried to define new `systemd_stdio_bridge_t` and `systemd_stdio_bridge_exec_t` types, label `/usr/bin/systemd-stdio-bridge` as `systemd_stdio_bridge_exec_t`, and then let only `systemd_stdio_bridge_t` be allowed to read/write `admindomain` `unix_stream_socket`s, but that didn't work, because [systemd runs `systemd-stdio-bridge` via `systemd-run`](https://github.com/systemd/systemd/blob/391ad5d8aa533010bed02079c95eab34de0408c5/src/libsystemd/sd-bus/sd-bus.c#L1614-L1622), and processes do not transition correctly when ran that way, so that solution is blocked by https://github.com/fedora-selinux/selinux-policy/issues/2500 and https://github.com/systemd/systemd/issues/32668.

So I've left the current functionality in, but I've changed it to use the proper interfaces now.

gucci-on-fleek committed:
Allow "snapper" to be ran by confined administrators

gucci-on-fleek commented:
(sorry for the long delay)

I've updated the code to use interfaces for all external types, and I've addressed most of the other comments. For the comments that I haven't addressed, I've replied with a few questions because I'm not sure how to proceed. And thanks again for the review!
